IQSS/dataverse
# Extend Datasets API to support dataset deaccession and recovery of files and file counts of deaccessioned datasets

**What this PR does / why we need it**:

This PR extends the following endpoints related to dataset version files:

- getVersionFiles ("{id}/versions/{versionId}/files")
- getVersionFileCounts ("{id}/versions/{versionId}/files/counts")

The above endpoints now accept a new boolean optional query parameter "includeDeaccessioned", which, if enabled, causes the endpoint to consider deaccessioned versions when searching for versions to obtain files or file counts. To do this, the following existing commands were modified:

- GetLatestAccessibleDatasetVersionCommand
- GetSpecificPublishedDatasetVersionCommand
- GetLatestPublishedDatasetVersionCommand

This logic is required for the SPA to display file metadata for deaccessioned dataset versions.

Additionally, a new endpoint has been developed to support version deaccessioning through API (Given a dataset and a version): "{id}/versions/{versionId}/deaccession".

The version identifiers ":latest", ":latest-published" and ":draft" have been extracted to API constants (ApiConstants.java), since these strings were hardcoded throughout the application and this way it is easier to manage them.

Finally, the DataFile API payload has been extended to add the field "friendlyType", as described in https://github.com/IQSS/dataverse-frontend/pull/169#issuecomment-1727018970

**Which issue(s) this PR closes**:

- Closes #9852

**Suggestions on how to test this**:

1) Create a dataset version and publish it

2) Deaccession dataset version via curl:

Create a deaccession.json file, which should like this:

```
{
  "deaccessionReason": "Description of the deaccession reason.",
  "deaccessionForwardURL": "https://demo.dataverse.org"
}

```
Export the following variables:

```
export API_TOKEN=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
export SERVER_URL=http://localhost:8080 (or your instance URL)
export ID=<DATASET_ID>
export VERSIONID=1.0
export FILE_PATH=deaccession.json
```

Then, call the deaccession endpoint:


`  curl -H "X-Dataverse-key:$API_TOKEN" -X POST "$SERVER_URL/api/datasets/$ID/versions/$VERSIONID/deaccession" -H "Content-type:application/json" --upload-file $FILE_PATH
`

3) Get version files for the deaccessioned dataset version:

`curl "$SERVER_URL/api/datasets/$ID/versions/$VERSIONID/files?includeDeaccessioned=true"`

4) Get version file counts for the deaccessioned dataset version:

`curl "$SERVER_URL/api/datasets/$ID/versions/$VERSIONID/files/counts?includeDeaccessioned=true"`

5) Repeat previous operations but removing includeDeaccessioned query param to check how the deaccessioned version is not found (old behavior).

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

No

**Is there a release notes update needed for this change?**:

Yes
